x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFeatures.java

```diff
@@ -0,0 +1,176 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+public class ConnectorFeatures implements Writeable, ToXContentObject {
+
+    private final boolean filteringAdvancedConfigEnabled;
+    private final boolean filteringRulesEnabled;
+    private final boolean incrementalSyncEnabled;
+    private final boolean syncRulesAdvancedEnabled;
+    private final boolean syncRulesBasicEnabled;
+
+    private ConnectorFeatures(
+        boolean filteringAdvancedConfig,
+        boolean filteringRules,
+        boolean incrementalSyncEnabled,
+        boolean syncRulesAdvancedEnabled,
+        boolean syncRulesBasicEnabled
+    ) {
+        this.filteringAdvancedConfigEnabled = filteringAdvancedConfig;
+        this.filteringRulesEnabled = filteringRules;
+        this.incrementalSyncEnabled = incrementalSyncEnabled;
+        this.syncRulesAdvancedEnabled = syncRulesAdvancedEnabled;
+        this.syncRulesBasicEnabled = syncRulesBasicEnabled;
+    }
+
+    public ConnectorFeatures(StreamInput in) throws IOException {
+        this.filteringAdvancedConfigEnabled = in.readBoolean();
+        this.filteringRulesEnabled = in.readBoolean();
+        this.incrementalSyncEnabled = in.readBoolean();
+        this.syncRulesAdvancedEnabled = in.readBoolean();
+        this.syncRulesBasicEnabled = in.readBoolean();
+    }
+
+    public static final ParseField FILTERING_ADVANCED_CONFIG_ENABLED_FIELD = new ParseField("filtering_advanced_config_enabled");
+    public static final ParseField FILTERING_RULES_ENABLED_FIELD = new ParseField("filtering_rules_enabled");
+    public static final ParseField INCREMENTAL_SYNC_ENABLED_FIELD = new ParseField("incremental_sync_enabled");
+    public static final ParseField SYNC_RULES_ADVANCED_ENABLED_FIELD = new ParseField("sync_rules_advanced_enabled");
+    public static final ParseField SYNC_RULES_BASIC_ENABLED_FIELD = new ParseField("sync_rules_basic_enabled");
+
+    private static final ConstructingObjectParser<ConnectorFeatures, Void> PARSER = new ConstructingObjectParser<>(
+        "connector_features",
+        true,
+        args -> new Builder().setFilteringAdvancedConfig((boolean) args[0])
+            .setFilteringRules((boolean) args[1])
+            .setIncrementalSyncEnabled((boolean) args[2])
+            .setSyncRulesAdvancedEnabled((boolean) args[3])
+            .setSyncRulesBasicEnabled((boolean) args[4])
+            .createConnectorFeatures()
+    );
+
+    static {
+        PARSER.declareBoolean(constructorArg(), FILTERING_ADVANCED_CONFIG_ENABLED_FIELD);
```

Comment:
This parser works fine processing the JSON body from and `updateFeatures` endpoint (that is yet to be implemented, but I tested it with put endpoint). The [elastic-connectors index template](https://github.com/elastic/elasticsearch/blob/b03939ec9d3fb90ac83ab62dcc3219cfbe41e536/x-pack/plugin/core/template-resources/src/main/resources/entsearch/connector/elastic-connectors-mappings.json#L33) forces us to write thr data to index as a complex, nested structure, see `toXContent` method in this class. 

I think the problematic case would be reading the features stored in an index and representing them as a `ConnectorFeatures` instance. In that case, we would need another parser that can work with deserialising XContent data stored in an index - but given its nested nature it's not straightforward to define for me. 

Am I correct to assume that this will only be an issue if we want to support partial updates to the individual fields?

Reply:
I'm not sure I understand correctly. Do you mean to have an external representation (API) that is different than the internal representation (index)?

I would advise to create separate classes and parsing methods for that. We do something similar in SearchApplications; storage is done [here](https://github.com/elastic/elasticsearch/blob/b7eafce32cb1e7dfdcb259aab89b886ff93a1c66/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java#L283) and read is done [here](https://github.com/elastic/elasticsearch/blob/b7eafce32cb1e7dfdcb259aab89b886ff93a1c66/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java#L428). It can be quite confusing as we do some binary parsing there, but the main point is we can use different classes for de/serialization

Reply:
> Do you mean to have an external representation (API) that is different than the internal representation (index)?

Yes, that's what I meant. And thank you for linking the examples :)